lib/ansible/module_utils/nxos.py

```diff
@@ -149,55 +144,41 @@ def run_commands(self, commands, check_rc=True):
             else:
                 cmd = item['command']
 
-            rc, out, err = self.exec_command(cmd)
+            out = connection.get(cmd)
             try:
                 out = to_text(out, errors='surrogate_or_strict')
             except UnicodeError:
                 self._module.fail_json(msg=u'Failed to decode output from %s: %s' % (cmd, to_text(out)))
 
-            if check_rc and rc != 0:
-                self._module.fail_json(msg=to_text(err))
-
-            if not check_rc and rc != 0:
-                try:
-                    out = self._module.from_json(err)
-                except ValueError:
-                    out = to_text(err).strip()
-            else:
-                try:
-                    out = self._module.from_json(out)
-                except ValueError:
-                    out = to_text(out).strip()
-
-            if item['output'] == 'json' and out != '' and isinstance(out, string_types):
-                self._module.fail_json(msg='failed to retrieve output of %s in json format' % item['command'])
+            try:
+                out = self._module.from_json(out)
+            except ValueError:
+                out = to_text(out).strip()
 
             responses.append(out)
         return responses
 
     def load_config(self, config, return_error=False, opts=None):
         """Sends configuration commands to the remote device
         """
-        if opts is None:
-            opts = {}
+        connection = self._get_connection()
+        connection.edit_config(config)
 
-        rc, out, err = self.exec_command('configure')
-        if rc != 0:
-            self._module.fail_json(msg='unable to enter configuration mode', output=to_text(err))
+    def get_capabilities(self):
+        connection = self._get_connection()
+        info = connection.get_capabilities()
 
-        msgs = []
-        for cmd in config:
-            rc, out, err = self.exec_command(cmd)
-            if opts.get('ignore_timeout') and rc == 1:
```

Comment:
@trishnaguha Just noticed that you removed the `ignore_timeout` flag that I had added as part of the nxos_install_os ISSU work.  This will likely break that functionality.  I need to take a closer look at this but we will likely need to look at how we can best maintain ISSU capability.

Reply:
@mikewiebe This is handled in the module nxos_install_os itself for now https://github.com/ansible/ansible/pull/31524/files#diff-fba4522179413344afa5de53c18542beR166
However there will be a commit in the PR soon so that we will be able to handle this scenario in module_utils level.

Reply:
Thanks @trishnaguha.  I ran our `nxos_install_os` tests against your changes and ran into an issue when transport is `nxapi`.  The tests appear to be ok for transport `cli`.

I did verify that both transports work fine in `devel`.

Here is the error:
```bash
TASK [nxos_install_os : Install OS image nxos.7.0.3.I7.1.bin] **********************************************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_install_os/tests/common/sanity.yaml:25
<n9k-109.cisco.com> connection transport is nxapi
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_install_os.py
<n9k-109.cisco.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n9k-109.cisco.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288 `" && echo ansible-tmp-1511983853.78-97350491039288="` echo $HOME/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288 `" ) && sleep 0'
<n9k-109.cisco.com> PUT /tmp/tmpdTEDF1 TO /root/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288/nxos_install_os.py
<n9k-109.cisco.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288/ /root/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288/nxos_install_os.py && sleep 0'
<n9k-109.cisco.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288/nxos_install_os.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1511983853.78-97350491039288/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py", line 598, in <module>
    main()
  File "/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py", line 583, in main
    install_result = do_install_all(module, issu, sif, kick=kif)
  File "/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py", line 505, in do_install_all
    impact_data = check_mode(module, issu, image, kick)
  File "/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py", line 494, in check_mode
    data = check_mode_nextgen(module, issu, image, kick)
  File "/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py", line 468, in check_mode_nextgen
    data = parse_show_install(load_config(module, commands, True, opts))
  File "/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py", line 171, in load_config
    rc, out, err = exec_command(module, 'configure')
  File "/tmp/ansible_NzWq9R/ansible_modlib.zip/ansible/module_utils/connection.py", line 66, in exec_command
  File "/tmp/ansible_NzWq9R/ansible_modlib.zip/ansible/module_utils/connection.py", line 99, in __init__
AssertionError: socket_path must be a value

fatal: [n9k-109.cisco.com]: FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py\", line 598, in <module>\n    main()\n  File \"/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py\", line 583, in main\n    install_result = do_install_all(module, issu, sif, kick=kif)\n  File \"/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py\", line 505, in do_install_all\n    impact_data = check_mode(module, issu, image, kick)\n  File \"/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py\", line 494, in check_mode\n    data = check_mode_nextgen(module, issu, image, kick)\n  File \"/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py\", line 468, in check_mode_nextgen\n    data = parse_show_install(load_config(module, commands, True, opts))\n  File \"/tmp/ansible_NzWq9R/ansible_module_nxos_install_os.py\", line 171, in load_config\n    rc, out, err = exec_command(module, 'configure')\n  File \"/tmp/ansible_NzWq9R/ansible_modlib.zip/ansible/module_utils/connection.py\", line 66, in exec_command\n  File \"/tmp/ansible_NzWq9R/ansible_modlib.zip/ansible/module_utils/connection.py\", line 99, in __init__\nAssertionError: socket_path must be a value\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
	to retry, use: --limit @/root/agents-ci/ansible/test/integration/nxos.retry

PLAY RECAP *************************************************************************************************************************************************************************************************************
n9k-109.cisco.com          : ok=26   changed=0    unreachable=0    failed=1   
```

Here is the playbook used for the test:
```yaml
---
- debug: msg="START TRANSPORT:{{ connection.transport }} nxos_os_install sanity test"

- set_fact: si='nxos.7.0.3.I6.1.bin'
#- set_fact: si='n3500-uk9.6.0.2.A8.6.bin'
#- set_fact: si='n3500-uk9.6.0.2.A8.8.bin'
#- set_fact: si='n7000-s2-dk9.8.0.1.bin'

#- set_fact: ki='n3500-uk9-kickstart.6.0.2.A8.6.bin'
#- set_fact: ki='n3500-uk9-kickstart.6.0.2.A8.8.bin'
#- set_fact: ki='n7000-s2-kickstart.8.0.1.bin'

- set_fact: checkmode='yes'

- set_fact: tv='7.0(3)I6(1)'
#- set_fact: tv='8.0(1)'
#- set_fact: tv='6.0(2)A8(8)'

- set_fact: issu='no'

- name: "Install OS image {{ si }}"
  check_mode: "{{ checkmode }}"
  nxos_install_os:
    system_image_file: "{{ si }}"
    #kickstart_image_file: "{{ ki }}"
    issu: "{{ issu }}"
    provider: "{{ connection | combine({'timeout': 500}) }}"
  register: result

- debug: msg=" {{ result }}"

- name: Wait for device to come back up with new image
  wait_for:
    port: 22
    state: started
    timeout: 500
    delay: 60
    host: "{{ inventory_hostname }}"
  when: "result.changed == true"

- name: "Check installed OS for newly installed version"
  nxos_command:
    commands: ['show version | json']
    provider: "{{ connection }}"
  register: output

- assert:
    that:
    - output['stdout'][0]['kickstart_ver_str'] == "{{ tv }}"
  when: not checkmode


- debug: msg="END TRANSPORT:{{ connection.transport }} nxos_os_install sanity test"
```

Reply:
@mikewiebe thanks for the test. this is WIP, will be fixed in the future commits to this PR.